src/analysis/ml.py

```diff
@@ -82,45 +91,66 @@ def pca(dataframe: pd.DataFrame, output_png: str, output_file: str, output_coord
     @param output_png: the filename to save the scatterplot
     @param output_file: the filename to save the variance explained by each component
     @param output_coord: the filename to save the coordinates of each algorithm
+    @param components: the number of principal components to calculate (Default is 2)
+    @param labels: determines if labels will be included in the scatterplot (Default is True)
     """
     df = dataframe.reset_index(drop=True)
+    columns = dataframe.columns
+    column_names = [element.split('-')[1] for element in columns]
     df = df.transpose()  # based on the algorithms rather than the edges
     X = df.values
 
+    min_shape = min(df.shape)
+    if components < 2:
+        raise ValueError(f"components={components} must be greater than or equal to 2 in the config file.")
+    elif components > min_shape:
+        print(f"components={components} is not valid. Setting components to {min_shape}.")
+        components = min_shape
+    if not isinstance(labels, bool):
+        raise ValueError(f"labels={labels} must be True or False")
+
     scaler = StandardScaler()
     scaler.fit(X)  # calc mean and standard deviation
     X_scaled = scaler.transform(X)
 
     # choosing the PCA
-    pca_2 = PCA(n_components=2)
-    pca_2.fit(X_scaled)
-    X_pca_2 = pca_2.transform(X_scaled)
-    variance = pca_2.explained_variance_ratio_ * 100
+    pca = PCA(n_components=components)
+    pca.fit(X_scaled)
+    X_pca = pca.transform(X_scaled)
+    variance = pca.explained_variance_ratio_ * 100
 
     # making the plot
+    label_color_map = create_palette(column_names)
     plt.figure(figsize=(10, 7))
-    sns.scatterplot(x=X_pca_2[:, 0], y=X_pca_2[:, 1], s=70)
+    sns.scatterplot(x=X_pca[:, 0], y=X_pca[:, 1], s=70, hue=column_names, legend=True, palette=label_color_map);
     plt.title("PCA")
     plt.xlabel(f"PC1 ({variance[0]:.1f}% variance)")
     plt.ylabel(f"PC2 ({variance[1]:.1f}% variance)")
 
-    # saving the PCA plot
-    make_required_dirs(output_png)
-    plt.savefig(output_png)
+    # saving the coordinates of each algorithm
+    make_required_dirs(output_coord)
+    data = {'algorithm': columns.tolist(), 'x': X_pca[:, 0], 'y': X_pca[:, 1]}
```

Comment:
If components is > 2, should we write all of the components to this file? Otherwise we are not using them.

Reply:
It would make sense to only give them the top 2, but there is a chance that giving the user all the components could be of value to them. It could help a user understand how much information is kept or lost based on the number of components they choose. Or it could also show them that for their use case, PCA could not be the best choice for using dimensionality reduction. 

We could either make it les confusing and show them the top two (which already happens in the graph), or we can show them the whole thing and hopefully it will aid them when using pca in spras.

Reply:
I think it would make the most sense to visualize the top 2 in the plot, which we already do, and then show the requested number of components in the written file here.

Reply:
I believe that is what it is doing

Reply:
The columns of `coord_df` should correspond to the number of components. We currently only create an `algorithm` column and then an `x` and `y` column for the first two components even if `X_pca` has larger dimension. We need to rename the column header so that instead of `x` and `y` is it more general like `comp1`, `comp2`, ... (or whatever is considered standard terminology, let's check the sklearn docs).

To confirm the behavior, I ran the EGFR config file with 4 components. The variance output file has 4 values:
```
60.39857079045826
20.81242547010328
10.34166118187788
8.447342557560589
```
but this coordinates file does not have 4 numeric columns:
```
algorithm	x	y
tps-egfr-pathlinker-params-7S4SLU6	5.169316618551315	-0.46981993977169956
tps-egfr-pathlinker-params-TCEMRS7	8.623647079828245	2.8919659890867155
tps-egfr-omicsintegrator1-params-ITO5EQS	-1.3678061290895147	-4.656992756839353
tps-egfr-omicsintegrator1-params-3THRXWW	-4.330653935914073	-2.8668877196329863
tps-egfr-omicsintegrator1-params-5QH767V	-8.094503633375975	5.1017344271573215
```